core/src/main/java/org/elasticsearch/index/IndexSortConfig.java

```diff
@@ -0,0 +1,218 @@
+            Function.identity(), Setting.Property.IndexScope, Setting.Property.Final);
+
+    /**
+     * The {@link SortOrder} for each specified sort field (ie. <b>asc</b> or <b>desc</b>).
+     */
+    public static final Setting<List<SortOrder>> INDEX_SORT_ORDER_SETTING =
+        Setting.listSetting("index.sort.order", Collections.emptyList(),
+            IndexSortConfig::parseOrderMode, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static SortOrder parseOrderMode(String value) {
+        try {
+            return SortOrder.fromString(value);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Illegal sort order:" + value);
+        }
+    }
+
+    /**
+     * The {@link MultiValueMode} for each specified sort field (ie. <b>max</b> or <b>min</b>).
+     */
+    public static final Setting<List<MultiValueMode>> INDEX_SORT_MODE_SETTING =
+        Setting.listSetting("index.sort.mode", Collections.emptyList(),
+            IndexSortConfig::parseMultiValueMode, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static MultiValueMode parseMultiValueMode(String value) {
+        MultiValueMode mode = MultiValueMode.fromString(value);
+        if (mode != MultiValueMode.MAX && mode != MultiValueMode.MIN) {
+            throw new IllegalArgumentException("Illegal index sort mode:[" + mode + "], " +
+                "must be one of [" + MultiValueMode.MAX + ", " + MultiValueMode.MIN + "]");
+        }
+        return mode;
+    }
+
+    /**
+     * The missing value for each specified sort field (ie. <b>_first</b> or <b>_last</b>)
+     */
+    public static final Setting<List<String>> INDEX_SORT_MISSING_SETTING =
+        Setting.listSetting("index.sort.missing", Collections.emptyList(),
+            IndexSortConfig::validateMissingValue, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static String validateMissingValue(String missing) {
+        if ("_last".equals(missing) == false && "_first".equals(missing) == false) {
+            throw new IllegalArgumentException("Illegal missing value:[" + missing + "], " +
+                "must be one of [_last, _first]");
+        }
+        return missing;
+    }
+
+    final String[] fields;
+    final SortOrder[] orders;
+    final MultiValueMode[] modes;
+    final String[] missingValues;
+
+    public IndexSortConfig(IndexSettings indexSettings) {
+        final Settings settings = indexSettings.getSettings();
+        if (INDEX_SORT_FIELD_SETTING.exists(settings)) {
+            fields = INDEX_SORT_FIELD_SETTING.get(settings)
+                .toArray(new String[0]);
+        } else {
+            fields = new String[0];
+        }
+        if (fields.length > 0 && indexSettings.getIndexVersionCreated().before(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            throw new IllegalArgumentException("unsupported index.version.created:" + indexSettings.getIndexVersionCreated() +
```

Comment:
How would we have gotten here? Would they need to use the test plugin to set the version? I'm not sure this is worth checking.